haskell/providers.bzl

```diff
@@ -218,3 +79,203 @@ GhcPluginInfo = provider(
+        dynamic_lib = None
+        if lib_to_link.dynamic_library:
+            dynamic_lib = lib_to_link.dynamic_library
+        elif lib_to_link.interface_library:
+            # XXX: Do these work with GHCi?
+            dynamic_lib = lib_to_link.interface_library
+        static_lib = None
+        if lib_to_link.pic_static_library:
+            static_lib = lib_to_link.pic_static_library
+        elif lib_to_link.static_library and hs.toolchain.is_windows:
+            # NOTE: GHCi cannot load non-PIC static libraries, except on Windows.
+            static_lib = lib_to_link.static_library
+
+        if dynamic_lib:
+            dynamic_lib = symlink_dynamic_library(hs, dynamic_lib, fixed_lib_dir)
+        static_lib = mangle_static_library(hs, dynamic_lib, static_lib, fixed_lib_dir)
+
+        lib = static_lib if static_lib else dynamic_lib
+        if dynamic and dynamic_lib:
+            lib = dynamic_lib
+
+        if lib:
+            libs.append(lib)
+
+    # NOTE: We can avoid constructing these in the future by instead generating
+    #   a dedicated package configuration file defining the required libraries.
+    sep = ";" if hs.toolchain.is_windows else None
+    library_path = make_path(libs, prefix = path_prefix, sep = sep)
+    ghc_env = {
+        "LIBRARY_PATH": library_path,
+        "LD_LIBRARY_PATH": library_path,
+    }
+
+    return (depset(direct = libs), ghc_env)
+
+def get_extra_libs(hs, dynamic, cc_info):
+    """Get libraries appropriate for linking with GHC.
+
+    GHC expects dynamic and static versions of the same library to have the
+    same library name. Static libraries for which this is not the case will be
+    symlinked to a matching name.
+
+    Furthermore, dynamic libraries will be symbolically linked into a common
+    directory to allow for less RPATH entries.
+
+    Args:
+      hs: Haskell context.
+      dynamic: Whether to prefer dynamic libraries.
+      cc_info: Combined CcInfo provider of dependencies.
+
+    Returns:
+      depset of File: the libraries that should be passed to GHC for linking.
+
+    """
+    fixed_lib_dir = target_unique_name(hs, "_libs")
+    libs_to_link = _get_unique_lib_files(cc_info)
+    static_libs = []
+    dynamic_libs = []
+    for lib_to_link in libs_to_link:
+        dynamic_lib = None
+        if lib_to_link.dynamic_library:
+            dynamic_lib = lib_to_link.dynamic_library
+        elif lib_to_link.interface_library:
+            dynamic_lib = lib_to_link.interface_library
+        static_lib = None
+        if lib_to_link.pic_static_library:
+            static_lib = lib_to_link.pic_static_library
+        elif lib_to_link.static_library:
+            static_lib = lib_to_link.static_library
+
+        if dynamic_lib:
+            dynamic_lib = symlink_dynamic_library(hs, dynamic_lib, fixed_lib_dir)
+        static_lib = mangle_static_library(hs, dynamic_lib, static_lib, fixed_lib_dir)
+
+        if dynamic and dynamic_lib:
```

Comment:
What does it mean to be `not (dynamic and dynamic_lib)`? In that case it gets added to `static_libs`, like when `static_lib != None`. Is that right?

Reply:
Yes, `dynamic` is a bool, `True` means prefer dynamic lib, `False` means prefer `static_lib`. We can only use `dynamic_lib` if it exists. That explains the first branch. Second branch, if there is no `static_lib` then we have to use the `dynamic_lib`. Third, default to `static_lib`. First and second could be merged into `if (dynamic and dynamic_lib) or not static_lib`, though I'm not sure if thats more readable.